pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	Shoot *Shoot
+	// Seed describes the Seed that will be registered.
+	// Either Seed or Gardenlet must be specified.
+	Seed *SeedTemplateSpec
+	// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+	Gardenlet *Gardenlet
+}
+
+// Shoot identifies the Shoot that will be registered as Seed.
+type Shoot struct {
+	// Name is the name of the Shoot that will be registered as Seed.
+	Name string
+	// APIServer specifies certain kube-apiserver parameters of the Shoot that will be registered as Seed.
+	APIServer *APIServer
+}
+
+// SeedTemplateSpec describes the data a Seed should have when created from a template.
+type SeedTemplateSpec struct {
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the desired behavior of the Seed.
+	Spec gardencorev1beta1.SeedSpec
+}
+
+// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+type Gardenlet struct {
+	// Deployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+	// the image, which bootstrap mechanism to use (bootstrap token / service account), etc.
+	Deployment *GardenletDeployment
+	// Config is the GardenletConfiguration used to configure gardenlet.
+	Config *gardenletconfigv1alpha1.GardenletConfiguration
+	// Bootstrap is the mechanism that should be used for bootstrapping gardenlet connection to the Garden cluster. One of ServiceAccount, Token.
+	// If specified, a service account or a bootstrap token will be created in the garden cluster and used to compute the bootstrap kubeconfig.
+	// If not specified, the gardenClientConnection.kubeconfig field will be used to connect to the Garden cluster.
+	GardenConnectionBootstrap *GardenConnectionBootstrap
+	// SeedConnection is the mechanism for gardenlet connection to the Seed cluster. Must equal ServiceAccount if specified.
+	// If not specified, the seedClientConnection.kubeconfig field will be used to connect to the Seed cluster.
+	SeedConnection *SeedConnection
```

Comment:
The point is, it's **not** the default in `use-as-seed`, see the long comment a few lines below this one. If we want to maintain "hard compatibility" (hopefully we don't, but still), we should really set this to `nil` meaning that we inherit whatever is in parent. Also, specifying a `kubeconfig` explicitly is a valid option for local debugging and in other scenarios where you don't care much about the security implications. 
And I think it doesn't belong to `Deployment`, it should be on the same level as `GardenConnectionBootstrap`. Currently `Deployment` only contains the values that are used to parameterize the gardenlet Helm chart and this is not one of them.

Reply:
> The point is, it's not the default in use-as-seed, see the long comment a few lines below this one. If we want to maintain "hard compatibility" (hopefully we don't, but still), we should really set this to nil meaning that we inherit whatever is in parent.

I didn't get that. The current behaviour reuses the SeedClientConnection of the parent gardenlet and this is by default a `ServiceAccount`: https://github.com/gardener/gardener/blob/789188822abcbd7c65d033aa5179a251426fcce9/pkg/gardenlet/controller/shoot/seed_registration_control.go#L702

> Also, specifying a `kubeconfig` explicitly is a valid option for local debugging and in other scenarios where you don't care much about the security implications.

How would one specify a kubeconfig in this field? I'm a bit confused.

> And I think it doesn't belong to `Deployment`, it should be on the same level as `GardenConnectionBootstrap`. Currently `Deployment` only contains the values that are used to parameterize the gardenlet Helm chart and this is not one of them.

Hmm, `seedClientConnection` is actually a deployment parameter of the gardenlet chart. It specifies, how the gardenlet 
should talk to its seed API server.

Reply:
@timebertt I understand that you are confused, let me try to explain. I struggled myself quite a bit with this.

I think you misunderstand a bit what `SeedConnection` means here and why it was introduced. If `SeedConnection` above is set to `ServiceAccount`, the controller forces using a service account by cleaning the `kubeconfig` field of the `seedClientConnection` if it is set. If it's `nil`, then it doesn't clean it, allowing for whatever is specified in the configuration to take effect. If `MergeWithParent` is `true` (as in `use-as-seed`), then whatever is set in the parent takes effect unless it's explicitly overwritten in the configuration (usually it's not).

> I didn't get that. The current behaviour reuses the SeedClientConnection of the parent gardenlet and this is by default a ServiceAccount

It doesn't matter what is the default in the chart. What matters is that whatever is in the parent is not touched by the old implementation, so it behaves as if the above setting is `nil`. This is the reason that if you try testing `use-as-seed` annotation with `no-gardenlet` in a local dev setup it will always fail, also with a `remote-garden`. To make it work (in order to check the compatibility with my implementation), I had to change the line you cite above to **not** propagate the `kubeconfig`, therefore forcing "service account".

> How would one specify a kubeconfig in this field? I'm a bit confused.

`kubeconfig` is a field of the `ClientConnectionConfiguration`. What I mean is that it **could** be valid (e.g. in a local dev setup you could just specify the kubeconfig of your shoot), so using a "service account" should be configurable. But I don't have a strong opinion here.

> Hmm, seedClientConnection is actually a deployment parameter of the gardenlet chart

Yes, but the seed connection "mode" in the sense `SeedConnection` is introduced above is not.

To make the long story short: if we don't care too much about compatibility (I also think we shouldn't care, but I am also a bit wary of breaking stuff), and if we are fine with limiting some local dev setup options (not critical), then we could remove the `SeedConnection` setting which means that "service account" will **always** be forced (and whatever is specified in the configuration / inherited from the parent lost).

Reply:
Ah, ok. I think, I now got your point.

One thing though: the `no-gardenlet` option can work in a local dev setup, but only in conjunction with `with-secret-ref`. This way, no gardenlet will be deployed, but a kubeconfig secret will be referenced in the SeedSpec, allowing the local gardenlet to connect to the shooted seed.

What I am still confused about is: when you use the `no-gardenlet` option, this field won't be set, because it belongs to the `Gardenlet` type, right?

However, this field is rather a minor detail. And if there is still some confusion/uncertainty, we can have a quick chat about it once you're back :)

Reply:
Sorry, I meant the `gardenlet` option (when `no-gardenlet` is not present). `no-gardenlet` works fine, the other one is broken in `use-as-seed` currently.

Reply:
@danielfoehrKn I understand your concerns with higher level field, I don't like them either, but here there is a bit of a dilemma that justifies introducing it I believe. Let me try to answer your questions:

> Why is that?  ...

* If `MergeParentConfig` is false, then you are right, we don't need a special setting since the user could simply choose to set `kubeconfig` or not. However, I assume this case is less "interesting" and will be used rarely, also the `use-as-seed` annotation implies `MergeParentConfig` to be true.
* If `MergeParentConfig` is true and `kubeconfig` is not set in the parent, then again we have no issue. I think this is the case in most production setup and this is the only reason the `use-as-seed` annotation works in production with the current implementation.
* If `MergeParentConfig` is true and `kubeconfig` is set in the parent, then it's practically guaranteed to be wrong and the resulting seed will be broken. Unfortunately, this is exactly the case in a local development setup, which makes `use-as-seed` untestable as of today. This is a huge drawback from my perspective, especially having in mind the work that is still pending on GEP-13.

So at least for `MergeParentConfig == true`, `seedClientConnection.kubeconfig` *must* be cleared, forcing "service account. I think having the `seedConnection` parameter as proposed is a reasonably good way to achieve it, without either forcing the "service account" or living with a broken local development setup.

Reply:
We decided to remove the `seedConnection` setting and make sure "service account" is always used, unless the user specified  `seedClientConnection.kubeconfig` in the `ManagedSeed` resource.